examples/example.cpp

```diff
@@ -115,138 +115,84 @@ class Node {
-}
 
-static const char *nodeAsString(const Uast* ctx, NodeHandle node) {
-  return ((Node *)node)->val_string.data();
-}
 
-static int64_t nodeAsInt(const Uast* ctx, NodeHandle node) {
-  return ((Node *)node)->val_int;
-}
+  NodeKind Kind() { return kind; }
 
-static uint64_t nodeAsUint(const Uast* ctx, NodeHandle node) {
-  return ((Node *)node)->val_uint;
-}
+  const char* AsString() { return val_string.data(); }
+  int64_t     AsInt()    { return val_int; }
+  uint64_t    AsUint()   { return val_uint; }
+  double      AsFloat()  { return val_float; }
+  bool        AsBool()   { return val_bool; }
 
-static double nodeAsFloat(const Uast* ctx, NodeHandle node) {
-  return ((Node *)node)->val_float;
-}
-
-static bool nodeAsBool(const Uast* ctx, NodeHandle node) {
-  return ((Node *)node)->val_bool;
-}
-
-static size_t nodeSize(const Uast* ctx, NodeHandle node) {
-  auto n = (Node *)node;
-  if (n->kind == NODE_ARRAY) {
-    return n->arr.size();
+  size_t Size() {
+    if (kind == NODE_ARRAY) {
+      return arr.size();
+    }
+    return obj.size();
   }
-  return n->obj.size();
-}
 
-static const char * nodeKeyAt(const Uast* ctx, NodeHandle node, size_t index) {
-  auto n = (Node *)node;
-  size_t i = 0;
-  for (auto it = n->obj.begin(); it != n->obj.end(); ++it, ++i) {
-    if (i == index) return it->first.data();
+  const char* KeyAt(size_t index) {
+    size_t i = 0;
+    for (auto it = obj.begin(); it != obj.end(); ++it, ++i) {
+      if (i == index) return it->first.data();
+    }
+    return NULL;
   }
-  return NULL;
-}
 
-static NodeHandle nodeValueAt(const Uast* ctx, NodeHandle node, size_t index) {
-  auto n = (Node *)node;
-  if (n->kind == NODE_ARRAY) {
-    if (index >= n->arr.size()) return 0;
-    return (NodeHandle)n->arr[index];
-  }
-  size_t i = 0;
-  for (auto it = n->obj.begin(); it != n->obj.end(); ++it, ++i) {
-    if (i == index) return (NodeHandle)it->second;
+  Node* ValueAt(size_t index) {
+    if (kind == NODE_ARRAY) {
+      if (index >= arr.size()) return 0;
+      return arr[index];
+    }
+    size_t i = 0;
+    for (auto it = obj.begin(); it != obj.end(); ++it, ++i) {
+      if (i == index) return it->second;
+    }
+    return NULL;
   }
-  return 0;
-}
-
-static NodeHandle nodeNewObject(const Uast* ctx, size_t size) {
-  auto n = new Node(NODE_OBJECT);
-  return (NodeHandle)(n);
-}
 
-static NodeHandle nodeNewArray(const Uast* ctx, size_t size) {
-  auto n = new Node(NODE_ARRAY);
-  n->arr.resize(size);
-  return (NodeHandle)(n);
-}
+  void SetValue(size_t i, Node* v) {
+    arr[i] = v;
+  }
 
-static NodeHandle nodeNewString(const Uast* ctx, const char * v) {
-  auto n = new Node(std::string(v));
-  return (NodeHandle)(n);
-}
+  void SetKeyValue(const char * k, Node* v) {
+    obj[k] = v;
+  }
+};
 
-static NodeHandle nodeNewInt(const Uast* ctx, int64_t v) {
-  auto n = new Node(v);
-  return (NodeHandle)(n);
-}
+class Creator : public uast::NodeCreator<Node*> {
+public:
+    Node* NewObject(size_t size) {
```

Comment:
I don't think it will work for this specific case. `Node*` will be converted to `NodeHandle` at some point, so we will lose reference counting anyway.